File: android/sample-ui/src/main/kotlin/com/usesmileid/sampleapps/ui/components/UseSmileIDSampleToast.kt
package com.usesmileid.sampleapps.ui.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.smileid.designsystem.SmileDimens
import com.usesmileid.sampleapps.ui.UseSmileIDSampleTestIds
import com.usesmileid.sampleapps.ui.theme.UseSmileIDSampleTheme

/**
 * A snackbar: a dark bar with a message and an underlined action, spanning the width it is given.
 *
 * Keeps the `sample_toast*` ids: the design node is named "toast", and renaming would churn four apps.
 */
@Composable
fun UseSmileIDSampleToast(
    message: String,
    modifier: Modifier = Modifier,
    actionLabel: String? = null,
    onAction: (() -> Unit)? = null,
) {
    Surface(
        modifier = modifier
            .fillMaxWidth()
            .defaultMinSize(minHeight = SNACKBAR_HEIGHT)
            .testTag(UseSmileIDSampleTestIds.TOAST),
        shape = RoundedCornerShape(SNACKBAR_RADIUS),
        color = UseSmileIDSampleTheme.colors.textTitle,
        shadowElevation = SNACKBAR_ELEVATION,
    ) {
        // FlowRow so a cramped action moves onto its own line whole; a Row breaks "Undo" in half.
        FlowRow(
            modifier = Modifier.padding(horizontal = SmileDimens.spacingMd, vertical = SNACKBAR_PADDING_Y),
            horizontalArrangement = Arrangement.spacedBy(SmileDimens.spacingSm),
            verticalArrangement = Arrangement.spacedBy(SmileDimens.space4),
            itemVerticalAlignment = Alignment.CenterVertically,
        ) {
            Text(
                text = message,
                style = UseSmileIDSampleTheme.type.bannerTextFont.copy(
                    fontSize = SNACKBAR_TEXT_SIZE,
                    fontWeight = FontWeight.Medium,
                ),
                color = UseSmileIDSampleTheme.colors.background,
                modifier = Modifier.weight(1f),
            )
            if (actionLabel != null && onAction != null) {
                Text(
                    text = actionLabel,
                    style = UseSmileIDSampleTheme.type.linkFont.copy(
                        fontSize = SNACKBAR_ACTION_SIZE,
                        fontWeight = FontWeight.SemiBold,
                        textDecoration = TextDecoration.Underline,
                    ),
                    color = UseSmileIDSampleTheme.colors.background,
                    softWrap = false,
                    textAlign = TextAlign.Center,
                    // Clickable before the sizing modifiers, so the tap target is the padded box.
                    // Widened, not squared off: a 48-high action inflates the 46 bar to 72.
                    modifier = Modifier
                        .testTag(UseSmileIDSampleTestIds.TOAST_UNDO)
                        .clickable(role = Role.Button, onClick = onAction)
                        .defaultMinSize(minWidth = SmileDimens.sizeControlMd)
                        .padding(horizontal = SmileDimens.spacingXs),
                )
            }
        }
    }
}

private val SNACKBAR_RADIUS = 12.dp
private val SNACKBAR_HEIGHT = 46.dp
private val SNACKBAR_PADDING_Y = 13.dp
private val SNACKBAR_TEXT_SIZE = 13.5.sp
private val SNACKBAR_ACTION_SIZE = 14.sp
private val SNACKBAR_ELEVATION = 10.dp

File: flutter/sample_ui/lib/src/components/use_smileid_sample_toast.dart
import 'package:flutter/material.dart';

import '../theme/use_smileid_sample_colors.dart';
import '../theme/use_smileid_sample_theme.dart';
import '../theme/use_smileid_sample_typography.dart';
import '../tokens/smile_tokens.dart';
import '../use_smileid_sample_test_ids.dart';

/// A snackbar: a dark bar with a message and an underlined action, spanning the width it is given.
///
/// Keeps the `sample_toast*` ids — the design node is named "toast", and renaming churns four apps.
class UseSmileIDSampleToast extends StatelessWidget {
  /// An action needs both halves; either one alone draws the message-only variant.
  const UseSmileIDSampleToast({
    required this.message,
    this.actionLabel,
    this.onAction,
    super.key,
  });

  /// The confirmation being reported.
  final String message;

  /// The action's label — "Undo" on a removal, "Make active" on a created profile.
  final String? actionLabel;

  /// What the action does; the offer is consumed on dismissal, so there is no second chance.
  final VoidCallback? onAction;

  @override
  Widget build(BuildContext context) {
    final UseSmileIDSampleColors colors = UseSmileIDSampleTheme.colorsOf(context);
    final String? label = actionLabel;
    final VoidCallback? action = onAction;
    final bool hasAction = label != null && action != null;

    return Semantics(
      identifier: UseSmileIDSampleTestIds.toast,
      container: true,
      child: Container(
        // Spans the width it is given, between the 16 margins the design draws it inside.
        width: double.infinity,
        decoration: BoxDecoration(
          color: colors.textTitle,
          borderRadius: const BorderRadius.all(Radius.circular(SmileDimens.radiusMd)),
          boxShadow: const <BoxShadow>[SmileShadows.elevationFloating],
        ),
        child: ConstrainedBox(
          constraints: const BoxConstraints(minHeight: _barHeight),
          child: Padding(
            padding: const EdgeInsets.symmetric(
              horizontal: SmileDimens.spacingMd,
              vertical: _barPaddingY,
            ),
            child: LayoutBuilder(
              builder: (BuildContext context, BoxConstraints constraints) => Wrap(
                spacing: SmileDimens.spacingSm,
                runSpacing: SmileDimens.space4,
                crossAxisAlignment: WrapCrossAlignment.center,
                children: <Widget>[
                  // Bounded rather than flexible: Wrap gives a child unbounded width, so without
                  // this the message never wraps, and with it a long one pushes the action onto
                  // its own line whole instead of breaking the word in half.
                  ConstrainedBox(
                    constraints: BoxConstraints(maxWidth: constraints.maxWidth),
                    child: Text(
                      message,
                      style: UseSmileIDSampleType.bannerTextFont.copyWith(
                        fontSize: _messageSize,
                        fontWeight: FontWeight.w500,
                        color: colors.background,
                      ),
                    ),
                  ),
                  if (hasAction) _action(colors, label, action),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }

  /// Widened, not squared off: a 48-high action inflates the 46 bar to 72, so the bar's own height
  /// is the vertical target and the minimum width carries the rest.
  Widget _action(UseSmileIDSampleColors colors, String label, VoidCallback action) => Semantics(
    identifier: UseSmileIDSampleTestIds.toastUndo,
    button: true,
    child: GestureDetector(
      onTap: action,
      behavior: HitTestBehavior.opaque,
      child: ConstrainedBox(
        constraints: const BoxConstraints(minWidth: SmileDimens.sizeControlMd),
        child: Padding(
          padding: const EdgeInsets.symmetric(horizontal: SmileDimens.spacingXs),
          child: Text(
            label,
            textAlign: TextAlign.center,
            softWrap: false,
            style: UseSmileIDSampleType.linkFont.copyWith(
              fontSize: _actionSize,
              fontWeight: FontWeight.w600,
              decoration: TextDecoration.underline,
              decorationColor: colors.background,
              color: colors.background,
            ),
          ),
        ),
      ),
    ),
  );
}

/// The design's own bar height; no scale token carries 46, and 48 would inflate the bar to 72.
const double _barHeight = 46;

/// The design's vertical padding, which no scale token carries.
const double _barPaddingY = 13;

/// The message run, a half-point off the nearest token.
const double _messageSize = 13.5;

/// The action run.
const double _actionSize = 14;
